### ISSUE TITLE
FSDP2 reduce_scatter_reduce_op for context parallelism

### ISSUE THREAD

**Issue by dingqingy** [pytorch/torchtitan#1132]
Hi,

FSDP2 reduce_scatter by default seems to take the average over the entire shard world, which consists of dp_shard and cp. Averaging gradients over dp_shard makes sense, but I wonder if sum is the better reduce op for CP?

Logically, it seems to me gradient should be agnostic to the choice of CP.

Thanks!

**Comment by fegin** [comment]
@dingqingy This is a good question. IMO, the average is the correct reduction. This is the loss function:

```
    return torch.nn.functional.cross_entropy(
        pred.flatten(0, 1).float(), labels.flatten(0, 1)
    )
```

As we flatten batch dimension and sequence dimension, and the default reduction of `cross_entropy` is `mean`, so both `dp_shard` (batch dimension) and `cp` (sequence dimension) should apply `mean` as the reduction.

**Comment by dingqingy** [comment]
@fegin Thanks for your response! Your explanation makes sense to me. Closing the issue.
